dashboard/tasks.py

```diff
@@ -0,0 +1,129 @@
+"""
+Periodic task that updates user data.
+"""
+import logging
+import uuid
+from datetime import datetime, timedelta
+
+from celery import group
+from django.conf import settings
+from django.contrib.auth.models import User
+from django.core.cache import cache
+
+from edx_api.client import EdxApi
+from social.apps.django_app.default.models import UserSocialAuth
+
+from backends import utils
+from backends.edxorg import EdxOrgOAuth2
+from dashboard.models import (
+    CachedCertificate,
+    CachedEnrollment,
+)
+from dashboard import api
+from micromasters.celery import async
+
+
+log = logging.getLogger(__name__)
+
+
+MAX_STUDENT_CHUNK_SIZE = 20  # max 20 students per sub task
+MAX_HRS_MAIN_TASK = 6
+PARALLEL_RATE_LIMIT = '10/m'
+LOCK_EXPIRE = 60 * 60 * 2  # Lock expires in 2 hrs
+lock_id = '{0}-lock'.format(uuid.uuid4())
+
+
+def chunks(students, chunk_size=MAX_STUDENT_CHUNK_SIZE):
+    """
+    Divid list into sub lists each of max size chunk_size.
+    Args:
+        students (List): list of studnets
+        chunk_size (Number): Max size of each sublist
+
+    Returns:
+    List: List of sublist containing student ids.
+    """
+    chunk_size = max(1, chunk_size)
+    for i in range(0, len(students), chunk_size):
+        yield students[i:i + chunk_size]
+
+
+@async.task
+def batch_update_user_data():
+    """
+    Create sub tasks to update user data like enrollments,
+    certificates and grades from edX platform.
+    """
+    def acquire_lock():
+        """
+        create lock by adding a key to storage.
+        """
+        # lock will expire if 2 hrs are pass and task is not complete.
+        return cache.add(lock_id, 'true', LOCK_EXPIRE)
+
+    def release_lock():
+        """
+        remove lock by deleteing key from storage.
+        """
+        return cache.delete(lock_id)
+
+    if acquire_lock():
+        enrolled_students = CachedEnrollment.objects.filter(
+            last_request__lt=datetime.utcnow() - timedelta(hours=MAX_HRS_MAIN_TASK)
+        ).values_list('user', flat=True)
+
+        certificated_students = CachedCertificate.objects.filter(
+            last_request__lt=datetime.utcnow() - timedelta(hours=MAX_HRS_MAIN_TASK)
+        ).values_list('user', flat=True)
+
+        students = list(set(enrolled_students) | set(certificated_students))
+
+        job = group(
+            batch_update_user_data_subtasks.s(list_students) for list_students in chunks(students)
+        )
+        result = job.apply_async()
+        result.ready()
+
+        if result.successful():
+            # release lock if all tasks are done.
+            release_lock()
```

Comment:
what happens if a child process fails for some reason? the lock is never released? Or am I misunderstanding the `result.successful()` call?